### ISSUE TITLE
syntax: newline style around &&, || and | configurable in the printer

### ISSUE THREAD

**Issue by legionus** [mvdan/sh#21]
Right now the code:

```
[ -z "$HOME" ] && foo || bar
[ -z "$HOME" ] &&
    foo ||
    bar
```

will be formatted:

```
[ -z "$HOME" ] && foo || bar
[ -z "$HOME" ] \
    && foo \
    || bar
```

Can you add an option or something to control the formatting of the line breaks ?

**Comment by mvdan** [comment]
The former style was actually used at first, but I decided in favour of the second since it is more consistent and more widely used, especially since it's used in the google shell style:

https://google.github.io/styleguide/shell.xml#Pipelines

`shfmt` should have as little options as possible, ideally none for its formatting. It currently only has one, concerning indentation. If we add an option for the indentation of pipelines, I fear we will soon have a dozen of tunables that aren't really necessary and overcomplicate things.

**Comment by legionus** [comment]
The google's favour is more cumbersome. The extra slashes make more noise. This is the main reason of my request.

Are you planning to implement google style described by reference ?

**Comment by mvdan** [comment]
> The google's favour is more cumbersome. The extra slashes make more noise. This is the main reason of my request.

That's true, but it makes for a more consistent and readable code. And most impostantly, it is more common.

Any command that spans multiple lines already uses backslashes, by piggybacking on `&&`, `||` and such you're just avoiding them. Being explicit about them is slightly more verbose but better IMHO, for the reasons stated above.

> Are you planning to implement google style described by reference?

The printer is not following the google style - it is too vague to define a full format. I'm simply quoting it as one of the reasons behind me saying "this is more popular".

**Comment by legionus** [comment]
I see.

**Comment by mvdan** [comment]
Multiple people have asked for this in the past year, so I'm giving it some more thought.

For reference, I decided to decline this some time ago because I wanted to avoid having lots of options in the printer and `shfmt`. But if it turns out that everyone is happy with just two options - indentation and this - that seems fair.

The default can't change, at least in 1.x. I don't think it should change after that either, as it's the only style that makes part of a good style guide (Google's, see above).

I don't know what the names of the config and flag could be - perhaps `PrintConfig.ShorterBinaryNewline` for the config (a bool) and `-shortnl` for `shfmt`. Suggestions welcome.

cc @uli42 @legionus

**Comment by mvdan** [comment]
Ah, and the person who also asked for this on Reddit was in fact @RalphCorderoy - CCing him too.

**Comment by RalphCorderoy** [comment]
To be clear when trying to suggest a name, this new option with choose between the time-honoured, Bell Labs, Bourne-shell-author, backslash-free style of `&&` and `||` as trailing _hyphens_ on a line indicating they're to continue, and not-known-for-their-sh-prowess Google's noisy `\`-`\n␣␣␣␣&&`? :-)

**Comment by mvdan** [comment]
Sorry, I don't understand. Are you asking what will be the default? Or proposing a better default?

I'm open to swapping the defaults in 2.0, but not decided yet.

**Comment by RalphCorderoy** [comment]
I'm want to ensure I understand what the new option will be doing to help suggest good names for it.

**Comment by legionus** [comment]
@mvdan can you describe in more detail what will the options that you want to add ? and based on that we can give them good names :)

in `indent(1)`:
```
-in, --indent-level n    Set indentation level to n spaces.
```

in `clang-format(1)`
```
IndentWidth     2
TabWidth:        8
UseTab:          Never
```

**Comment by legionus** [comment]
As for me, there's nothing wrong with having more options to customize formatting and style to have google as the default settings.

**Comment by mvdan** [comment]
Okay, let me rephrase in more detail.

We're talking about options in the printer in the syntax package, really. Right now there is only one option:

https://godoc.org/github.com/mvdan/sh/syntax#PrintConfig

So indentation is already dealt with. `shfmt` exposes this via `-i` - see `shfmt -h`.

So, to make the style of binary commands (`&&`, `||` and `|`) with newlines configurable, we need to add a new field there (probably a boolean) and add a command-line flag to `shfmt` to expose it in the tool.

We can either keep the default as `foo \\\n    && bar` or change it to `foo &&\n    bar`. I don't have a strong opinion either way. It can't change in 1.x. I'm open to set the default to the non-backslash one in 2.0, but I'd like to see proof that it really is the more popular style or at least the original one in sh/bash.

Depending on what we decide the default is, we will need a config option and cli flag to switch it around. So I guess we need to decide that first before we think of names.

For the sake of not breaking things in 2.0, I'll vote for keeping the current default. But again, no strong feeling about it and happy to be convinced otherwise.

**Comment by legionus** [comment]
> We can either keep the default as foo \\\n && bar or change it to foo &&\n bar. I don't have a strong opinion either way. It can't change in 1.x. I'm open to set the default to the non-backslash one in 2.0, but I'd like to see proof that it really is the more popular style or at least the original one in sh/bash.

I can show you code from some projects, but it would still be my subjective assessment. But if you add the option, it will not matter what the default is.

@mvdan Maybe you misunderstood me... I never asked to change the default values. I only ask the possibility to change these values.

**Comment by mvdan** [comment]
> I can show you code from some projects, but it would still be my subjective assessment.

If from your experience the shorter form is way more common, that's still something to take into account.

> Maybe you misunderstood me... I never asked to change the default values.

I realise that - but if we are to add a library option and a cli flag, these need to be set on stone. So I want to make sure I get the default (and the names, later) right.

**Comment by mvdan** [comment]
Actually, does the printer need to enforce either? I just realised it could simply keep the current style, depending on whether or not the operand is on the first or second line.

I guess the only disadvantage then is that you can't enforce it - you would have to do it manually. Perhaps this is a better option than adding an option and a flag.

**Comment by legionus** [comment]
Yes. I want to enforce conversion from
```
[ -z "$HOME" ] \
    && foo \
    || bar

[ -z "$HOME" ]  && foo ||
    bar
```
and other forms to
```
[ -z "$HOME" ] &&
    foo ||
    bar
```
that's why I want an option.

**Comment by legionus** [comment]
Git mostly uses ending `&&` and `||`:
https://github.com/git/git/blob/35f6318d44379452d8d33e880d8df0267b4a0cd0/git-bisect.sh#L443
https://github.com/git/git/blob/35f6318d44379452d8d33e880d8df0267b4a0cd0/git-parse-remote.sh#L22

https://github.com/altlinux/hasher fully written in posix shell and follows same policy.

Unfortunately I realized that it's very difficult to identify projects where there is a strict policy on shell scripts.

**Comment by mvdan** [comment]
Point noted about enforcing of one of the styles. I would want the same.

Git is a good example, actually. This makes me think it's more widespread than I thought (especially since the maintainer there works at Google, so he doens't follow their shell style?)

And I just realised that the use of backslashes to join lines also means the formatter has to do weird things with comments, since `foo # bar \` doesn't join with the next line - see #21.

All in all, convinced myself that the default should be no backslashes and we should have options to use the longer (google) form. Unfortunately, this will have to wait for 2.0. On the plus side, I'm already considering 2.0 for other backwards incompatible changes so I'll make master be 2.x soon enough.

On to the names - since the names of the option and the flag will be to use the longer form with a backslash, they should say something like "binary operators in next line" or "binary operators at start of line". Perhaps `BinaryNextLine` and `-bn`.

**Comment by uli42** [comment]
How about "implicitWrap" for "no backslash"?

Vom Smartphone gesendet.

----- Ursprüngliche Nachricht -----
Von: "Daniel Martí" <notifications@github.com>
Gesendet: ‎21.‎04.‎2017 18:08
An: "mvdan/sh" <sh@noreply.github.com>
Cc: "Ulrich Sibiller" <uli42@gmx.de>; "Mention" <mention@noreply.github.com>
Betreff: Re: [mvdan/sh] syntax: newline style around &&, || and | configurablein the printer (#21)

Point noted about enforcing of one of the styles. I would want the same.
Git is a good example, actually. This makes me think it's more widespread than I thought (especially since the maintainer there works at Google, so he doens't follow their shell style?)
And I just realised that the use of backslashes to join lines also means the formatter has to do weird things with comments, since foo # bar \ doesn't join with the next line - see #21.
All in all, convinced myself that the default should be no backslashes and we should have options to use the longer (google) form. Unfortunately, this will have to wait for 2.0. On the plus side, I'm already considering 2.0 for other backwards incompatible changes so I'll make master be 2.x soon enough.
On to the names - since the names of the option and the flag will be to use the longer form with a backslash, they should say something like "binary operators in next line" or "binary operators at start of line". Perhaps BinaryNextLine and -bn.

**Comment by legionus** [comment]
@mvdan I have an idea and I would like to gently (and very carefully) suggest it to you. In order not to invent many options for each parameter, you can make a config file and then the option will be only one: `-config FILE`. Of course, if you're not against configs :)

**Comment by mvdan** [comment]
> How about "implicitWrap" for "no backslash"?

It would need to have a word in there like `Binary`, `Pipe` or even `Chain` to make it obvious what it's doing. Otherwise the term `Wrap` is too ambiguous.

> I have an idea and I would like to gently (and very carefully) suggest it to you.

This scares me...

> In order not to invent many options for each parameter, you can make a config file and then the option will be only one: -config FILE. Of course, if you're not against configs :)

I assume that by "invent many options" you mean command-line flags for `shfmt`.

I'm not against configs per se. I've considered a config for `shfmt` before but decided against it because the amount of settings it currently has are very small. And adding config logic to the tool just seems overkill.

After all, if what you want is to not have to repeat the flags, you can just set up an alias or a shell script.

**Comment by legionus** [comment]
> I assume that by "invent many options" you mean command-line flags for `shfmt`.

Yep.

> And adding config logic to the tool just seems overkill.

Yes and no. You can unmarshal yaml to fill fields in the `PrintConfig`.

> After all, if what you want is to not have to repeat the flags, you can just set up an alias or a shell script.

I'm perfectly ok with command line options. I was just thinking that if in the future you decide to add more options the code that processes them will grow. I suggested config file only on the basis of this.

**Comment by mvdan** [comment]
> I'm perfectly ok with command line options. I was just thinking that if in the future you decide to add more options the code that processes them will grow. I suggested config file only on the basis of this.

It's true that reading a config file would be a bunch of lines with json/yaml, but using flags is just as easy - it takes three lines per flag. So unless it gets to the point where it has two dozen flags (hopefully not!), I don't think a config would make it any easier for me.

**Comment by RalphCorderoy** [comment]
Hi @mvdan, If you're still in doubt what the original style was then I can dig some examples up, but I've been writing Bourne-shell scripts for over 25 years and `\\\n&&` has only really appeared in the last five or so. I suspect it's carried over from another language that doesn't need the line-ending escaped, e.g. C++, though I find it looks wrong there too; not K&R. :-)

sh was clearly designed so the line could be split after `|`, etc., without a backslash, it's not a happy accident, and that was known and widely used.

So assuming 2.0 switches, I'm struggling to come up with a good name. I think it affects all binary operators?

    // Escape the line ending and put the binary operator at
    // the start of the next line. Affects "|", "|&", "&&", and "||".
    BackslashBinaryOp bool

As for _shfmt_(1), alternatives: `-b` for backslash, `-\` for, again, backslash, or `-g` for Google-style, in case more warts crop up. I can't think of a better case for a `-\` option to a command; it would need escaping from the shell, obviously. :-)

**Comment by mvdan** [comment]
> sh was clearly designed so the line could be split after |, etc., without a backslash, it's not a happy accident, and that was known and widely used.

Fair enough - that will be the default starting in 2.0.

> `BackslashBinaryOp bool`

Hmm, I can't decide between this and my `BinaryNextLine` now.

> Google-style, in case more warts crop up

I have no intent to add more printer options, and the Google style isn't explicit enough to dictate all of the printer. So I'm not considering this at all.

> `-\`

That would be a confusing one - rather not.

> `-b`

I want to use at least two chars, since this is a somewhat complicated option and I feel like a single char isn't enough. With my version I proposed `-bn` - with yours it could be `-bb`, for `BackslashBinary`.

**Comment by RalphCorderoy** [comment]
Ah, two characters, fair enough. I guess **-\\|** is out of the question. ;-) Yes, **-bb** seems good.

**Comment by mvdan** [comment]
I want to get 1.3.0 out soon and then I'll branch 1.x off, making master the 2.x branch. Once I do that I'll be able to work on this. I'll let you guys know once it's ready to test.

Thanks again for the input!

**Comment by uli42** [comment]
`BackslashBinary` might be technically correct. But we are talking about a command line switch which is made for humans. Typically people using shfmt just want to have a decent output and are not interested in the gory implementation or specification details. So from a user's perspective this switch treats line wraps. IMHO this should be reflected in the option name.

**Comment by mvdan** [comment]
Care to provide a name? Mine talked about the next line without mentioning the backslash - is that what you mean?

**Comment by uli42** [comment]
Well, how about something like "--alternativelinewrap"? I image some user not getting the result he expects and looking if there's an option that tweaks the programs behaviour. He might not be interested what exactly is different, just that line wraps are treated in another way.

> It would need to have a word in there like Binary, Pipe or even Chain to make it obvious what it's doing. 

I mean all those terms are not of interest for the ordinary user because they require too much knowledge. Everything the user sees are lines that are wrapped ugly. Most users know the term "pipe" but I doubt that too many ordinary users know what a "binary operator" is (really, I know lots of shell "programmers" that are just hacking until it works. They don't care to use the right terms or even to know them.)

**Comment by RalphCorderoy** [comment]
Just "line wrap" doesn't work because there's many cases of line-wrapping problems a user might come investigating and this is just one specific case where binary operators are "backslashed" and put on the next line. If the 1.0 user comes looking for that in 2.0 then the option name will make sense because they're writing sh scripts and can understand _binary operator_. If they want their long strings split and indented over many lines then _alternativelinewrap_ is vague enough to be a possible, but wrong, hit.

**Comment by mvdan** [comment]
I agree that "line wrap" is too generic. People might think it affects the position of `then` and such.

I guess `Binary` is not noob-friendly, but I can't think of a better term. "Pipe" is technically wrong as that's just half of the operators.

**Comment by legionus** [comment]
> I guess Binary is not noob-friendly, but I can't think of a better term. "Pipe" is technically wrong as that's just half of the operators.

`OperatorLastInLine` ?
`WrapByOperator` ?

**Comment by mvdan** [comment]
"operator" has the same problem as "line wrap" IMO - it's too generic.

I'm currently leaning towards my suggestion above, `BinaryNextLine` and `-bn`. Even if it's confusing, `-bn` will have a help text in `shfmt -h`.

**Comment by mvdan** [comment]
See the commits above - this is now implemented. Please build `shfmt` from master and let me know how it works.

`shfmt -bn` should behave the same as the older `shfmt`. The new `shfmt` should now enforce binary operators on the same line, and it should get rid of escaped newlines right before binary operators.

**Comment by mvdan** [comment]
@legionus @RalphCorderoy @uli42 did any of you have a chance to try it out? Would like it to be properly tested before I do a 2.0 release, in case there are any obvious bugs I missed or any breaking changes I won't be able to do later.

Thanks!

**Comment by uli42** [comment]
is on my todo list, but I lack time ATM

**Comment by RalphCorderoy** [comment]
Hi @mvdan, I've had a play. Feeding it

    false || # &&
    true \
    && \
    # ||
    echo end

without the new -bn gives

    false || # &&
        true &&

        # ||
        echo end

The blank line is surprising. Removing the " \" from the third line still gives the same output. But on "normal" input, it's fine and a lot nicer, thanks.

**Comment by mvdan** [comment]
Ah, the blank line is "expected" because only the statement lines are used to know whether or not to leave a blank line, not the operator lines. I should fix this. Mind opening an issue?
